File: src/EditorFeatures/VisualBasicTest/Structure/RegionDirectiveStructureTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports Microsoft.CodeAnalysis.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Outlining
    <Trait(Traits.Feature, Traits.Features.Outlining)>
    Public Class RegionDirectiveStructureProviderTests
        Inherits AbstractVisualBasicSyntaxNodeStructureProviderTests(Of RegionDirectiveTriviaSyntax)

        Friend Overrides Function CreateProvider() As AbstractSyntaxStructureProvider
            Return New RegionDirectiveStructureProvider()
        End Function

        <Fact>
        Public Async Function BrokenRegion() As Task
            Const code = "
$$#Region ""Goo""
"

            Await VerifyNoBlockSpansAsync(code)
        End Function

        <Fact>
        Public Async Function SimpleRegion() As Task
            Const code = "
{|span:$$#Region ""Goo""
#End Region|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Goo", autoCollapse:=False, isDefaultCollapsed:=True))
        End Function

        <Theory, CombinatorialData>
        Public Async Function RegionsShouldBeCollapsedByDefault(collapseRegionsByDefault As Boolean) As Task
            Const code = "
{|span:$$#Region ""Goo""
#End Region|}
"

            Dim options = New BlockStructureOptions() With {.CollapseRegionsWhenFirstOpened = collapseRegionsByDefault}

            Await VerifyBlockSpansAsync(code, options,
                Region("span", "Goo", autoCollapse:=False, isDefaultCollapsed:=collapseRegionsByDefault))
        End Function

        <Fact>
        Public Async Function RegionWithNoBanner1() As Task
            Const code = "
{|span:$$#Region
#End Region|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "#Region", autoCollapse:=False, isDefaultCollapsed:=True))
        End Function

        <Fact>
        Public Async Function RegionWithNoBanner2() As Task
            Const code = "
{|span:$$#Region """"
#End Region|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "#Region", autoCollapse:=False, isDefaultCollapsed:=True))
        End Function

        <Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/537984")>
        Public Async Function RegionEndOfFile() As Task
            Const code = "
Class C
End Class
{|span:$$#Region
#End Region|}"

            Await VerifyBlockSpansAsync(code,
                Region("span", "#Region", autoCollapse:=False, isDefaultCollapsed:=True))
        End Function
    End Class
End Namespace

File: src/EditorFeatures/CSharpTest/Structure/RegionDirectiveStructureTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CSharp.Structure;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Structure;
using Microsoft.CodeAnalysis.Test.Utilities;
using Roslyn.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Structure;

[Trait(Traits.Feature, Traits.Features.Outlining)]
public sealed class RegionDirectiveStructureTests : AbstractCSharpSyntaxNodeStructureTests<RegionDirectiveTriviaSyntax>
{
    internal override AbstractSyntaxStructureProvider CreateProvider() => new RegionDirectiveStructureProvider();

    [Fact]
    public Task BrokenRegion()
        => VerifyNoBlockSpansAsync("""
                $$#region Goo
                """);

    [Fact]
    public Task SimpleRegion()
        => VerifyBlockSpansAsync("""
                {|span:$$#region Goo
                #endregion|}
                """,
            Region("span", "Goo", autoCollapse: false, isDefaultCollapsed: true));

    [Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/539361")]
    public Task RegressionFor5284()
        => VerifyBlockSpansAsync("""
                namespace BasicGenerateFromUsage
                {
                    class BasicGenerateFromUsage
                    {
                        {|span:#reg$$ion TaoRegion

                        static void Main(string[] args)
                        {
                            /*Marker1*/
                            CustomStack s = new CustomStack(); //Generate new class

                            //Generate constructor
                            Classic cc = new Classic(5, 6, 7);/*Marker2*/

                            Classic cc = new Classic();
                            //generate property
                            cc.NewProperty = 5; /*Marker3*/

                        }
                        #endregion TaoRegion|}
                    }

                    class Classic
                    {
                    }
                }
                """,
            Region("span", "TaoRegion", autoCollapse: false, isDefaultCollapsed: true));

    [Theory, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/953668"), CombinatorialData]
    public async Task RegionsShouldBeCollapsedByDefault(bool collapseRegionsWhenFirstOpened)
    {
        var options = GetDefaultOptions() with
        {
            CollapseRegionsWhenFirstOpened = collapseRegionsWhenFirstOpened
        };

        await VerifyBlockSpansAsync("""
                class C
                {
                    {|span:#region Re$$gion
                    static void Main(string[] args)
                    {
                    }
                    #endregion|}
                }
                """, options,
            Region("span", "Region", autoCollapse: false, isDefaultCollapsed: collapseRegionsWhenFirstOpened));
    }

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/4105")]
    public Task SpacesBetweenPoundAndRegionShouldNotAffectBanner()
        => VerifyBlockSpansAsync("""
                class C
                {
                {|span:#  region R$$egion
                    static void Main(string[] args)
                    {
                    }
                #  endregion|}
                }
                """,
            Region("span", "Region", autoCollapse: false, isDefaultCollapsed: true));

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75583")]
    public Task Trailing()
        => VerifyBlockSpansAsync("""
            {|span:#region R$$1
            /* comment */ #endregion
            /* comment */ #region R2
            #endregion|}
            """,
            Region("span", "R1", autoCollapse: false, isDefaultCollapsed: true));
}
